// doxjs/js/index.js
import { DoxObject } from "./DoxObject";
import { isObject, warn, hasProperty, _toString, isFunction } from "./utils";
Array.prototype.forEach = function (callback) {
    for (let i = 0; i < this.length; i++) {
        callback.call(this, this[i], i, this);
    }
};
export default class DoxJS {
    constructor(target) {
        this.loneListeners = [];
        this._isRelyAnyVar = false;
        if (!isObject(target)) {
            warn(`DoxJS only accept object as argument`);
        }
        else {
            this.sourceObject = new DoxObject(target, this);
        }
    }
    execLonely() {
        this.loneListeners.forEach(f => f());
    }
    subscribe(listener, excuteLater = false, context) {
        if (isFunction(listener)) {
            this.runtimeEvent = {
                isLegal: true,
                value: listener
            };
            let fn = () => {
                // check if rely any var
                this._isRelyAnyVar = false;
                this.sourceObject.extendsRuntimeEvent(this.runtimeEvent);
                listener(this.sourceObject.observe());
                this.runtimeEvent.isLegal = false;
                // not rely any var
                if (!this._isRelyAnyVar) {
                    this.loneListeners.push(listener);
                }
                this._isRelyAnyVar = false;
            };
            return (excuteLater ? fn : fn());
        }
        else if (_toString.call(listener) === '[object String]') {
            if (hasProperty(this.listeners, listener)) {
                if (context) {
                    this.subscribe(this.listeners[listener].bind(context), excuteLater);
                }
                else {
                    this.subscribe(this.listeners[listener], excuteLater);
                }
            }
        }
        else {
            warn(`First parameter of subscribe should be Function or String`);
        }
    }
    bindListeners(listeners = {}) {
        this.listeners = listeners;
    }
    observe() {
        return this.sourceObject.observe();
    }
    bindActions(actions = {}) {
        this.actions = actions;
    }
    dispatch(action, ...args) {
        if (hasProperty(this.actions, action)) {
            this.actions[action](this.sourceObject.observe(), ...args);
        }
    }
}
